### PR TITLE
refactor: paragraph and block attribute

### PR DESCRIPTION
**Pull request by ldez**
Whole changes of the capture for block attribute .
Introducing paragraph patterns.

Support for:
- source
  - [x] paragraph
  - [x] listing block
  - [x] open block
- admonition
  - [x] paragraph
  - [x] inline
  - [x] example block
  - [x] open block
- quote
  - [x] quote block
  - [x] air quote (partial)
  - [x] markdown style
- passthrough
   - [x] paragraph
   - [x] passthrough block
   - [x] open block
   - [x] inline

Related to #125 #126

### CODE REVIEW

**Comment by ldez** [comment]
Support for:
- source
  - [x] paragraph
  - [x] listing block
  - [x] open block
- admonition
  - [x] paragraph
  - [x] inline
  - [x] example block
  - [x] open block
- quote
  - [x] quote block
  - [x] air quote (partial)
  - [x] markdown style

**Comment by ldez** [comment]
Complete rewrite of 'block attribute' pattern.
Complete rewrite for 'source', 'quote', 'admonition' patterns.

**Comment by ldez** [comment]
It's the first step to paragraph patterns.
The full rewrite with paragraph patterns come in an another PR after this PR will be merged

**Comment by ldez** [comment]
quotes;

capture du 2016-05-15 21-47-02

**Comment by ldez** [comment]
source:

capture du 2016-05-15 21-49-18

**Comment by ldez** [comment]
admonition:

capture du 2016-05-15 21-46-11

**Comment by ldez** [comment]
@mojavelinux @nicorikken any remarks?

File: spec/code-block-generator-spec.coffee [hunk 1]
```diff
@@ -51,66 +64,87 @@ describe 'Code block generator', ->
...
-        end: '(?<=----)[\\r\\n]+$'
+        end: '((?<=--)[\\r\\n]+$|^\\p{Blank}*$)'
 
     it 'should generate C++ code block', ->
       languages = [
         pattern: 'c(pp|\\+\\+)', type: 'source', code: 'cpp'
       ]
       codeBlocks = generator.makeAsciidocBlocks(languages)
       expect(codeBlocks).toHaveLength 3 # Number of blocks
+      # begin: '^\\[(source),\\p{Blank}*(?i:(c(pp|\\+\\+)))(?:,([^\]]*))?\\]$'
```
Comment (nicorikken) [inline spec/code-block-generator-spec.coffee]: I assume this comment line can be removed, as it seems to be more of a WIP annotation.

**Comment by nicorikken** [comment]
I spotted one comment left in the code changes, addressed with an inline comment (and you fixed while I was typing this message 😄 ) I've tried to break your features (based on the examples here), but failed to do so 😢 , so great work 👍. This level of nesting now introduced seems complicated, but I'm pretty sure I'll get the hang of it when I get into developing mode again.
**very nice work!** 👏

**Comment by nicorikken** [comment]
Ready, check, so merge!

**Comment by mojavelinux** [comment]
This is really unbelievable, @ldez (but I totally believe it). This grammar goes so far beyond any grammar that is out there for AsciiDoc. I actually think this grammar is now going to play an important role in defining a specification for AsciiDoc and, eventually, UniDoc. In fact, I'd even say that together, we might be able to solve the inline parsing problem in Asciidoctor core (wink wink). https://github.com/asciidoctor/asciidoctor/issues/61

**Comment by mojavelinux** [comment]
I just played around with the new package and discovered a case that may have been missed. I can add an ID to a quote paragraph as follows:

```
[quote#someone-quote]
This is just a quote from someone.
```

However, I cannot assign only a role:

```
[quote.famous]
This is just a quote from someone.
```

However, I can add a role if I also add an ID:

```
[quote#someone-quote.famous]
This is just a quote from someone.
```

The following pattern should be allowed:

```
[quote#id]
[quote.role]
[quote%option]
[quote.role#id]
[quote#id.role]
[quote%option#id]
[quote.role%option]
etc...
```

Notice that order doesn't matter. ID can only be specified once, but role and option can be specified any number of times.

**Comment by mojavelinux** [comment]
To start, we may not have to worry about validating the content of what comes after the block name (e.g., quote) as long as `#`, `.` and `%` are allowed.

**Comment by ldez** [comment]
Open an issue, I will try to solve this.

**Comment by ldez** [comment]
this examples already works:

```adoc
[quote#someone-quote]
This is just a quote from someone.
```
```adoc
[quote#someone-quote.famous]
This is just a quote from someone.
```

**Comment by ldez** [comment]
Fixed!!

**Pull request merged**